### DISCUSSION TITLE
How to mock a StateNotifier with Mockito or test StateNotifierProvider

### DISCUSSION THREAD

**Discussion by hongfeiyang** [rrousselGit/riverpod#1398] [Q&A]
**Is your feature request related to a problem? Please describe.**

I see there are ways to override providers like `StateProvider` and `Provider` with a mocked value and we can use `mockito` to create a mock so that we can verify if a function of the mock class has been called or not, but what about `StateNotifierProvider`? I tried to use a `mockito` generated mocks for a `StateNotifier`, and override a `StateNotifierProvider` with this value but it does not seems to work because definitely under the hood there are functions like `addListeners` that cannot be stubbed. 

A typical use case would be having a button in a screen that calls `ref.read(anyStateNotifierProvider.notifier).func()` and I would like to verify if `func()` in `anyStateNotifierProvider` has been called or not with `mockito`.

So my question is, what is the recommended way to use mockito with StateNotifierProvider or StateNotifier?

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by rrousselGit** [comment]
You can do:

```dart
class MyNotifier extends StateNotifier<int> {
  ...
}

class MyNotifierMock extends StateNotifier<int> with Mock implements MyNotifier {
  MyNotifierMock(int initialValue): super(initialValue);
}
```

This should allow you to override the provider with a mock

**Reply by Felypee** [reply]
Maybe is a dumb question but why do you extend StateNotifier in the mocked class if it is already extended in MyNotifier which is being at the same time implemented in the MyNotifierMock?

**Reply by rrousselGit** [reply]
It's not extending MyNotifier, it's implementing it

So methods of StateNotifier like addListener/hasListeners & co are not mocked. But methods that MyNotifier defines are mocked

**Reply by kmcgill88** [reply]
Maybe another dumb question. I use `@GenerateMocks`, do I include `MyNotifierMock` to be generated and use that or just use `MyNotifierMock` directly?

**Reply by Plinzen** [reply]
@kmcgill88 : Did you find a solution to your question? I am also interested in a solution using `@GenerateMocks`. Thank you!

**Reply by samyakkkk** [reply]
The solution doesn't seem to work with `flutter_riverpod: ^2.3.10`
```dart
class MockAlbumsNotifer extends StateNotifier
    with Mock
    implements AlbumsNotifier {}
```
where AlbumsNotifer is:
```dart
class AlbumsNotifier extends StateNotifier<List<Album>?> {
  AlbumsNotifier(super.state, {required this.albumRepository});

  final AlbumRepository albumRepository;

  Future<void> loadAlbums() async {
    final albums = await albumRepository.fetchAlbums();
    state = albums;
  }
}
```
gives multiple `invalid_implementation_override` issues:
```
'StateNotifier.addListener' ('void Function() Function(void Function(dynamic), {bool fireImmediately})') isn't a valid concrete implementation of 'StateNotifier.addListener' ('void Function() Function(void Function(List<Album>?), {bool fireImmediately})').dart[invalid_implementation_override](https://dart.dev/diagnostics/invalid_implementation_override)

'StateNotifier.stream' ('Stream<dynamic> Function()') isn't a valid concrete implementation of 'StateNotifier.stream' ('Stream<List<Album>?> Function()').dart[invalid_implementation_override](https://dart.dev/diagnostics/invalid_implementation_override)

'StateNotifier.state' ('dynamic Function()') isn't a valid concrete implementation of 'StateNotifier.state' ('List<Album>? Function()').dart[invalid_implementation_override](https://dart.dev/diagnostics/invalid_implementation_override)

// and so on
```

Need more help here since we've an app that where we're trying on a project with StateNotifer based architecture and are unable to write widget tests. The only way seems to be faking.

**Reply by rrousselGit** [reply]
You need to specify the generics on StateNotifier

**Reply by samyakkkk** [reply]
Solved! thanks.

**Reply by tyukesz** [reply]
This solution works, but for the sake of conventions I would like to have every mock generated, which is the way suggested by Mockito. 
Or at least isn't there a way to mock the ref object itself? When I'm unit testing my widgets, I don't really care what is inside the provider. It would be nice if I could just mock the `ref.read` and `ref.watch`, therefore I could test the conditional renderings based on `ref.watch` return value and also verify if `ref.read` was called on gesture events.
